tests/register/register.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package register
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/coreos/init/tests/util"
+)
+
+type Test struct {
+	Name            string
+	Func            func(*testing.T, Test)
+	BinaryPath      string
+	IgnitionConfig  *string
+	CloudinitConfig *string
+}
+
+func (test Test) Run(t *testing.T) {
+	if os.Getenv("TMPDIR") == "" {
+		tmpDir, err := ioutil.TempDir("/var/tmp", "")
+		if err != nil {
+			t.Fatalf("failed to create temp working dir in /var/tmp: %v", err)
+		}
+
+		err = os.Setenv("TMPDIR", tmpDir)
+		if err != nil {
+			t.Fatalf("couldn't set TMPDIR env var: %v", err)
+		}
+
+		defer test.RemoveAll(t, tmpDir)
+		defer os.Setenv("TMPDIR", "")
+	}
+	test.Func(t, test)
+}
+
+func (test Test) CreateDevice(t *testing.T) (string, string) {
+	diskFile, err := ioutil.TempFile("", "coreos-install-disk")
```

Comment:
You'll want to avoid `TempFile` here; just use `OpenFile` with a fixed filename in `os.TempDir()`.  The file descriptor you're leaking will prevent the OS from freeing the space used by the file until the process exits, so once you have a large number of tests, you'll be consuming a significant amount of space in `/var/tmp`.

Really: once you have an outer tmpdir per test, you don't need any further TempFile calls at all; fixed names are okay.

Reply:
Switching them to directly named files.

However, isn't the leak solved by having a defer'd RemoveAll on the TMPDIR at the end of each individual test (inside of the `test.Run` method) when TMPDIR isn't specified?

I'm adding defer'd `os.Remove` calls on the files inside of TMPDIR in the case that it was not empty in `test.Run`.

Reply:
> However, isn't the leak solved by having a defer'd RemoveAll on the TMPDIR at the end of each individual test (inside of the test.Run method) when TMPDIR isn't specified?

RemoveAll is necessary but not sufficient.  The OS will not free the blocks used by a file until all of the file's directory entries have been deleted _and_ all of its file descriptors have been closed.

> I'm adding defer'd os.Remove calls on the files inside of TMPDIR in the case that it was not empty in test.Run.

Whoops, I missed that case.  A better approach would be to continue creating per-test temporary directories, but inside TMPDIR if that was specified.

Reply:
Just pushed another change which changes the approach in `Test.Run` to follow this as well as adding defer'd file.Close calls.